### ISSUE TITLE
Problems using fit_generator with a custom generator

### ISSUE THREAD

**Issue by singlasahil14** [fchollet/keras#5284]
```
f = h5py.File(path+'results/precomp.h5', 'r')
conv_trn_feat = f['train_features'][:]
trn_labels = np.tile(f['train_labels'][:], reps=(2,1))

def myGenerator():
    while True:
        for i in range(0, len(conv_trn_feat), batch_size):
            yield conv_trn_feat[i:i+batch_size], trn_labels[i:i+batch_size]

train_datagen = myGenerator()

bn_model = Sequential(get_bn_layers(0, num=num_hidden))
bn_model.compile(Adam(lr=0.001), loss='categorical_crossentropy', metrics=['accuracy'])
bn_model.optimizer.lr = 1e-3
bn_model.fit_generator(train_datagen, samples_per_epoch=len(conv_trn_feat), nb_epoch=4)
bn_model = Sequential(get_bn_layers(0, num=num_hidden))
bn_model.compile(Adam(lr=0.001), loss='categorical_crossentropy', metrics=['accuracy'])
bn_model.optimizer.lr = 1e-3
bn_model.fit(conv_trn_feat, trn_labels, nb_epoch=4)`

```

I am precomputing the features of a convolutional layer and training fully connected layers using them. I have to use hdf5 file and a custom generator because the size of dataset is quite large.

Following is the output:
<img width="544" alt="screen shot 2017-02-06 at 11 36 13 am" src="https://cloud.githubusercontent.com/assets/6660192/22636432/0c2d05a4-ec61-11e6-9193-1a8bd0ac64df.png">

**The loss when using fit decreases much more quickly then using fit_generator with a custom generator.**
Am I doing something wrong? To me it looks like some bug in fit_generator.

**Comment by singlasahil14** [comment]
I had to shuffle the batches. That did the trick.
